src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp

```diff
@@ -361,6 +1121,87 @@ void dfcc_instrumentt::instrument_function_body(
 
   auto &body = goto_function.body;
 
+  // Builds a dependency graph representing the loop nesting structure
+  // identifies loop head and creates fresh symbols for write sets
+  grapht<loop_graph_nodet> loop_nesting_graph;
+  compute_loop_nesting_graph(function_id, loop_nesting_graph, body);
+
+  local_may_aliast local_may_alias(goto_function);
+  // Iterate over the (natural) loops in the function, in topo-sorted order,
+  // and apply any loop contracts that we find.
+  for(const auto &idx : loop_nesting_graph.topsort())
```

Comment:
Ideally our loop dependency graph should be acyclic, but we could have bugs that end up making it cyclic.
In case the graph is cyclic `topsort` will return an empty sequence. As sanity check we should check that the topsort has as many nodes as we added to the graph (it proves acyclicity).

Reply:
Yes, we will add the check.